Mods/DiscordRichPresence/Src/DiscordRpc.cpp

```diff
@@ -0,0 +1,203 @@
+#include "DiscordRpc.h"
+
+#include "Hooks.h"
+#include "Logging.h"
+#include "DiscordClient.h"
+
+#include <Glacier/ZScene.h>
+#include <regex>
+
+DiscordClient discordClient;
+
+DiscordRpc::~DiscordRpc()
+{
+	discordClient.Teardown();
+}
+
+void DiscordRpc::PreInit()
+{
+	discordClient.Initialize();
+	PopulateScenes();
+	PopulateGameModes();
+	PopulateCodenameHints();
+	Hooks::ZEntitySceneContext_LoadScene->AddDetour(this, &DiscordRpc::OnLoadScene);
+}
+
+void DiscordRpc::PopulateScenes()
+{
+	m_scenes = {
+		// Menu
+		{ "boot.entity", "In Startup Screen" },
+		{ "mainmenu.entity", "In Menu" },
+		// Sniper Assassin
+		{ "hawk", "Himmelstein" },
+		{ "salty", "Hantu Port" },
+		{ "caged", "Siberia" },
+		// Prologue
+		{ "thefacility", "ICA Facility" },
+		// HITMAN
+		{ "paris", "Paris" },
+		{ "coastaltown", "Sapienza" },
+		{ "marrakesh", "Marrakesh" },
+		{ "bangkok", "Bangkok" },
+		{ "colorado", "Colorado" },
+		{ "hokkaido", "Hokkaido" },
+		// HITMAN 2
+		{ "sheep", "Hawke's Bay" },
+		{ "miami", "Miami" },
+		{ "colombia", "Santa Fortuna" },
+		{ "mumbai", "Mumbai" },
+		{ "skunk", "Whittleton Creek" },
+		{ "theark", "Isle of Sgàil" },
+		{ "greedy", "New York" },
+		{ "opulent", "Haven Island" },
+		// HITMAN 3
+		{ "golden", "Dubai" },
+		{ "ancestral", "Dartmoor" },
+		{ "edgy", "Berlin" },
+		{ "wet", "Chongqing" },
+		{ "elegant", "Mendoza" },
+		{ "trapped", "Carpathian Mountains" }
+	};
+	Logger::Info("Finished populating scene map");
+}
+
+void DiscordRpc::PopulateGameModes()
+{
+	m_gameModes = {
+		{ "sniper", "Sniper Assassin" },
+		{ "usercreated", "Contracts Mode" },
+		{ "creation", "Contracts Mode" },
+		{ "featured", "Featured Contract" },
+		{ "mission", "Mission" },
+		{ "flashback", "Mission" },
+		{ "tutorial", "Mission" },
+		{ "campaign", "Mission" },
+		{ "escalation", "Escalation" },
+		{ "elusive", "Elusive Target" }
+	};
+	Logger::Info("Finished populating game modes");
```

Comment:
Avoid printing stuff with `Info` level that are not relevant to the user. These should be printed with `Debug` or even `Trace`.